File: src/bin/day07.py
import os
import sys
from typing import Self

class Node:
    least_size = sys.maxsize

    def __init__(self, size, children, parent):
        self.size = size
        self.children = children
        self.parent = parent
    
    def get_total_size(self):
        if not self.children:
            return self.size
        
        return sum([node.get_total_size() for node in self.children.values()])
    
    @classmethod
    def get_least_size(cls, node: Self, space_needed: int):
        size = node.get_total_size()
        if size >= space_needed and size < cls.least_size:
            cls.least_size = size
        if not node.children:
            return
        for node in node.children.values():
            cls.get_least_size(node, space_needed)

cur_node = Node(size=0, children={}, parent=None)
root = cur_node
    
with open(os.path.join(os.path.dirname(__file__), "..\\..\\inputs\\day7.txt")) as f:
    lines = f.readlines()
    i = 1
    while i < len(lines):
        line = lines[i]
        parts = line.split()
        cmd = parts[1]
        if cmd == "cd":
            if parts[2] == "..":
                cur_node = cur_node.parent
            else:
                cur_node = cur_node.children[parts[2]]
            i += 1
        elif cmd == "ls":
            i += 1
            while i < len(lines) and not lines[i].startswith("$"):
                output = lines[i].split()
                if output[0] == "dir":
                    node = Node(0, children={}, parent=cur_node)
                else:
                    node = Node(int(output[0]), children=None, parent=cur_node)
                cur_node.children[output[1]] = node
                
                i += 1

space_needed = 30000000 - (70000000 - root.get_total_size())
Node.get_least_size(root, space_needed)
print(Node.least_size)

File: src/bin/day07.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::rc::{Rc, Weak};

#[derive(Clone, Debug)]
enum Node {
    File(u64),
    Dir {
        children: HashMap<String, Rc<RefCell<Node>>>,
        parent: Option<Rc<RefCell<Node>>>,
    },
}

impl Node {
    // fn add_child(&mut self, name: String, child: Rc<RefCell<Node>>) {
    //     match self {
    //         Node::Dir {
    //             ref mut children,
    //             parent,
    //         } => {
    //             children.insert(name, child);
    //             println!("{:?}", children);
    //         }
    //         Node::File(_) => {
    //             panic!("Cannot call add_child on a file")
    //         }
    //     }
    // }
    fn get_size(&self) -> u64 {
        match self {
            Node::File(size) => *size,
            Node::Dir { children, .. } => {
                children.values().map(|node| node.borrow().get_size()).sum()
            }
        }
    }
    fn get_total_at_most_100k(&self) -> u64 {
        match self {
            Node::File(_) => 0,
            Node::Dir { children, .. } => {
                let size = self.get_size();
                return if size < 100000 { size } else { 0 }
                    + children
                        .values()
                        .map(|node| node.borrow().get_total_at_most_100k())
                        .sum::<u64>();
            }
        }
    }

    fn get_least_needed(&self, need: u64) -> Option<u64> {
        match self {
            Node::File(_) => None,
            Node::Dir { children, .. } => {
                let min_needed_size = children
                    .values()
                    .filter_map(|node| node.borrow().get_least_needed(need))
                    .min();

                if min_needed_size.is_some() {
                    return min_needed_size;
                }

                let size = self.get_size();
                if size >= need {
                    return Some(size);
                }

                None
            }
        }
    }
}

fn main() {
    let file = File::open("inputs/day7.txt").unwrap();
    let reader = BufReader::new(file);

    let root = Rc::new(RefCell::new(Node::Dir {
        children: HashMap::new(),
        parent: None,
    }));

    let mut cur_node = Rc::clone(&root);

    for line in reader.lines().skip(1) {
        if let Ok(line) = line {
            let line: Vec<&str> = line.split_whitespace().collect();

            match line.as_slice() {
                ["$", "ls"] => {}
                ["$", "cd", to] => match *to {
                    ".." => {
                        let parent = match cur_node.borrow().clone() {
                            Node::Dir { parent, .. } => parent,
                            _ => panic!("a"),
                        };
                        cur_node = parent.unwrap();
                    }
                    dir => {
                        let child = match cur_node.borrow().clone() {
                            Node::Dir { children, .. } => Rc::clone(children.get(dir).unwrap()),
                            _ => panic!("a"),
                        };
                        cur_node = child;
                    }
                },
                [s, filename] => match *s {
                    "dir" => {
                        if let Node::Dir {
                            ref mut children, ..
                        } = *cur_node.borrow_mut()
                        {
                            children.insert(
                                String::from(*filename),
                                Rc::new(RefCell::new(Node::Dir {
                                    children: HashMap::new(),
                                    parent: Some(Rc::clone(&cur_node)),
                                })),
                            );
                        }
                    }
                    size => {
                        if let Node::Dir {
                            ref mut children, ..
                        } = *cur_node.borrow_mut()
                        {
                            children.insert(
                                String::from(*filename),
                                Rc::new(RefCell::new(Node::File(size.parse::<u64>().unwrap()))),
                            );
                        }
                    }
                },
                _ => {}
            }
        }
    }
    println!(
        "Solution to part 1: {}",
        root.borrow().get_total_at_most_100k()
    );

    let needed = 30000000 - (70000000 - root.borrow().get_size());

    println!(
        "Solution to part 2: {}",
        root.borrow().get_least_needed(needed).unwrap()
    );
}
